judge/admin/contest.py

```diff
@@ -211,13 +216,45 @@ def make_hidden(self, request, queryset):
                                              count) % count)
     make_hidden.short_description = _('Mark contests as hidden')
 
+    def set_locked(self, request, queryset):
+        for row in queryset:
+            self.set_is_locked(row, True)
+        count = queryset.count()
+        self.message_user(request, ungettext('%d contest successfully locked.',
+                                             '%d contests successfully locked.',
+                                             count) % count)
+    set_locked.short_description = _('Lock contest submissions')
+
+    def set_unlocked(self, request, queryset):
+        for row in queryset:
+            self.set_is_locked(row, False)
+        count = queryset.count()
+        self.message_user(request, ungettext('%d contest successfully unlocked.',
+                                             '%d contests successfully unlocked.',
+                                             count) % count)
+    set_unlocked.short_description = _('Unlock contest submissions')
+
+    def set_is_locked(self, contest, is_locked):
+        with transaction.atomic():
+            contest.is_locked = is_locked
+            contest.save()
+            ContestSubmission.objects.filter(participation__contest=contest).update(submission__is_locked=is_locked)
```

Comment:
Would it be more efficient to use a `Submission.objects.filter(contest_object=contest).update(is_locked=True)` here instead?